### PR TITLE
Native fetch

### PR DESCRIPTION
**Pull request by daun**
**Description**

Next try of #569 after the rewrite to hooks.

- [x] Switch to native `fetch` API
- [x] Create a new hook `fetchPage` that plugins can hook before to return a fetch response
- [x] Create a new hook `loadPage` that plugins can hook before to return a page object
- [x] Create a new error type `FetchError` to identify fetch errors in loadPage
- [x] Test if server errors (500) are propagated properly as before
- [ ] Test this with plugins (forms and preload plugins specifically)
    - [ ] Make sure none of the plugins depend on the previously returned `xmlHttpRequest` directly
- [ ]  ~~Allow aborting requests~~ — will do later as part of `timeout` feature

**Plugin hooks**

This is how the preload plugin will now inject preloaded pages:

```js
this.swup.hooks.before('loadPage', (context, args) => {
  args.page = this.preloadPromises.get(args.url);
});
```

**Checks**

- [x] The PR is submitted to the `next` branch
- [x] The code was linted before pushing (`npm run lint`)
- [x] All tests are passing (`npm run test`)
- [x] New or updated tests are included
- [ ] ~~The documentation was updated as required~~

**Closes**

#566

### CODE REVIEW

**Comment by hirasso** [comment]
Aborting requests came up to gracefully handle hung up requests

**Comment by daun** [comment]
@hirasso Was that in relation to the planned `timeout` feature? I also remember us trying to make the preload promises abortable, but it's all been a while and my memory is hazy 😶‍🌫️

**Comment by hirasso** [comment]
Exactly, some kind of timeout was the plan IIRC :)

File: cypress/e2e/main.spec.js [hunk 1]
```diff
@@ -27,6 +27,57 @@ describe('Request', function () {
...
+	it('should force-load on network error', function () {
+		cy.intercept('/error-network.html', { times: 1 }, { forceNetworkError: true });
+		cy.shouldHaveReloadedAfterAction(() => {
+			this.swup.loadPage('/error-network.html');
+		});
+		cy.shouldBeAtPage('/error-network.html');
+	});
+});
+
+describe.only('Fetch', function () {
```
Comment (hirasso) [inline cypress/e2e/main.spec.js:48]: `.only` loves to sneak into PRs :)

File: src/helpers/Location.ts [hunk 2]
```diff
@@ -5,30 +5,33 @@
...
 	/**
 	 * Instantiate a Location from an element's href attribute
-	 * @param {Element} el
+	 * @param el
 	 * @return new Location instance
 	 */
-	static fromElement(el: HTMLAnchorElement): Location {
+	static fromElement(el: Element): Location {
 		const href = el.getAttribute('href') || el.getAttribute('xlink:href');
 		return new Location(href!);
```
Comment (hirasso) [inline src/helpers/Location.ts:26]: Maybe a drive-by improvement could be to make sure that the href is always a string here:
```js
const href = el.getAttribute('href') || el.getAttribute('xlink:href') || '';
return new Location(href);
```

File: src/Swup.ts [hunk 3]
```diff
@@ -332,21 +332,21 @@ export default class Swup {
...
-			popstate: true
-		});
 
 		// Does this even do anything?
 		// if (!hash) {
 		// 	event.preventDefault();
 		// }
 
 		this.hooks.triggerSync('popState', { event }, () => {
+			this.context = this.createContext({
```
Comment (hirasso) [inline src/Swup.ts:342]: Could you briefly explain why the context must be created inside the handler instead of before? Just for me to get a better feel for the concept.

File: src/modules/Hooks.ts [hunk 4]
```diff
@@ -16,17 +16,18 @@ export interface HookDefinitions {
 	clickLink: { event: DelegateEvent<MouseEvent> };
 	disabled: undefined;
 	enabled: undefined;
+	fetchPage: { url: string; options: FetchOptions; response?: Response | Promise<Response> };
+	loadPage: { url: string; options: FetchOptions; page?: PageData | Promise<PageData> };
```
Comment (hirasso) [inline src/modules/Hooks.ts:20]: Are you still planning to refactor `loadPage`, `performPageLoad` et al?

File: src/modules/fetchPage.ts [hunk 5]
```diff
@@ -1,54 +1,81 @@
...
+		this.name = 'FetchError';
+		this.url = details.url;
+		this.status = details.status;
+	}
+}
+
+/**
+ * Fetch a page from the server, return it and cache it.
+ */
+
```
Comment (hirasso) [inline src/modules/fetchPage.ts:29]: an empty line

File: src/modules/fetchPage.ts [hunk 6]
```diff
@@ -1,54 +1,81 @@
...
-			if (requestURL === url) {
-				this.cache.set(url, page);
-			}
-
-			this.hooks.trigger('pageLoaded', { page });
-			resolve(page);
-		});
-	});
+	// Resolve real url after potential redirect
+	const { url: finalUrl } = new Location(responseUrl);
```
Comment (hirasso) [inline src/modules/fetchPage.ts:68]: I keep being irritated by the naming of `Location`. I always think it's the `window.location` or something, even if that of course doesn't make sense. Anyways, a bit late to rename it now, isn't it? 🐌 `SwupLocation` or `SwupURL` would have been more clear IMHO.

**Review by hirasso — COMMENTED** [review]
Nice work! It's way to much to test without actually using it. Looking forward to a first release candidate! 🥳

File: cypress/e2e/main.spec.js (hunk 1 above)
Comment (daun) [reply cypress/e2e/main.spec.js:48]: 🫠

File: src/Swup.ts (hunk 3 above)
Comment (daun) [reply src/Swup.ts:342]: To avoid creating context in case the `popState` hook is replaced and the visit is never performed. But now that you mention it, I think it's better to revert this and make sure we're creating the context correctly, even if somebody replaces the hook.

File: src/modules/Hooks.ts (hunk 4 above)
Comment (daun) [reply src/modules/Hooks.ts:20]: Yes, I'd like to do but in another PR. This one is getting a bit too large.

File: src/modules/fetchPage.ts (hunk 6 above)
Comment (daun) [reply src/modules/fetchPage.ts:68]: I think `Location` is perfect. While `window.location` is the very specific `Location` the window is currently at, our `Location`s are free-floating ones that aren't attached to anything at the moment, just pointing at things.

For what it's worth, turbo calls this a `Locatable` which is slightly more weird but kind of the same thing.
